Add album-cover related gradients

GalexY727 opened the pull request:
This is most definitely a personal change, much like #23, but hey! I like colors and others might too. It does add a new dependency of colorthief, which is worth mentioning. Here's a gif preview :)
![Gif example](https://github.com/novatorem/novatorem/assets/65139378/699ef7e4-bb05-4ba1-906a-7e131c903533)


(It pulls colors from the album cover to make a palette for the song title, artist name, and eq bars)

GalexY727 committed:
Visualize light theme

GalexY727 committed:
sync with gradients branch

most likely will revert

GalexY727 committed:
Sorry light mode I am too tired for you

GalexY727 committed:
Merge branch 'novatorem:master' into master

GalexY727 committed:
Hope to add average color for the gradient

GalexY727 committed:
nth child for all bar in bars real bar foobar bar!

GalexY727 committed:
Make the EQ bar colors be based upon the album cover

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing into HEAD

GalexY727 committed:
Fix req.txt

GalexY727 committed:
silly imports, why are you so... you

GalexY727 committed:
Give the song title a nice gradient

GalexY727 committed:
textPalette -> songPalette

GalexY727 committed:
Dark mode compatibility

GalexY727 committed:
prevent .env changes from being made to the docker image, unless running locally

GalexY727 committed:
Add gradient to the text in the song name, attempt animation

GalexY727 committed:
remove animation from text

GalexY727 committed:
frantic commenting

GalexY727 committed:
Why does this work on a contianer from my computer but not on vercel

GalexY727 committed:
aha!

GalexY727 committed:
Add in some nice animations for the text colors

GalexY727 committed:
Add in some nice animations for the colors

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Add glow keyframes for the text

GalexY727 committed:
remove glow animation

GalexY727 committed:
remove text shadow entirely

GalexY727 committed:
force commit to rebuild vercel?

GalexY727 committed:
undo last change,(literally nothing)

GalexY727 committed:
Speed up artist gradient

novatorem commented on api/spotify.py:
```diff
@@ -111,6 +117,9 @@ def makeSVG(data, background_color, border_color):
     barCount = 84
     contentBar = "".join(["<div class='bar'></div>" for _ in range(barCount)])
     barCSS = barGen(barCount)
+    barPalette = gradientGen(data["item"]["album"]["images"][1]["url"], 4)
```
These two lines would error out when the user isn't actively listening to anything.
Probably best to make use of the two checks below (if playing + if album art exists)

```
[ERROR] Exception on /api/spotify [GET]
Traceback (most recent call last):
  File "/var/task/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/var/task/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/var/task/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/var/task/flask/_compat.py", line 39, in reraise
    raise value
  File "/var/task/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/var/task/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "./api/spotify.py", line 175, in catch_all
    svg = makeSVG(data, background_color, border_color)
  File "./api/spotify.py", line 120, in makeSVG
    barPalette = gradientGen(data["item"]["album"]["images"][1]["url"], 4)
KeyError: 'item'
```

GalexY727 replied:
Fixed, sorry about that one, I noticed it today

GalexY727 replied:
I added a PLACEHOLDER_URL as a band-aid to that, I completely overlooked those two checks :eyes:

GalexY727 committed:
Only set the barPalette if there is no song currently playing

Attempt to set songPalette to the recently playing url, but there may be a requirement for a barPalette

GalexY727 committed:
Give barPalette some love

GalexY727 committed:
Add placeholder URL if no image exists

GalexY727 commented:
It would be honestly better to use the placeholder image that is in the code there, but I am not 100% sure if colorthief will recognize b64

GalexY727 committed:
Enable personalized changes

GalexY727 committed:
Enable personalized changes

GalexY727 committed:
Personalized changes

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 committed:
Merge branch 'master' of https://github.com/GalexY727/spotify-now-playing

GalexY727 commented:
Feel free to omit changes since commit [836b7b4](https://github.com/novatorem/novatorem/pull/66/commits/836b7b4f9069eed5b0ba8f72eb840d56e189063c)

Edit: Those commit messages are horrifying... I-- I apologize.
All I did was enable the status message, make that text gradient (of course), then hide EQ bars when not listening to music

novatorem commented:
> Feel free to omit changes since commit [836b7b4](https://github.com/novatorem/novatorem/pull/66/commits/836b7b4f9069eed5b0ba8f72eb840d56e189063c)
> 
> Edit: Those commit messages are horrifying... I-- I apologize. All I did was enable the status message, make that text gradient (of course), then hide EQ bars when not listening to music

That's cool, it'll be a squash merge. Don't want to play around with cherry-picking a range nor rebasing, so I'll roll out your personalized changes.

Thank you for the work involved!
